pkg/csidriver/driver.go

```diff
@@ -244,38 +245,43 @@ func commonUpsertRanger(dv *driverVolume, key, value interface{}) error {
 	// TODO: prometheus metrics/alerts may be desired here, though some due diligence on what k8s level metrics/alerts
 	// around host filesystem issues might already exist would be warranted with such an exploration/effort
 
-	// Next, on an update we first nuke any existing directory and then recreate it to simplify handling the case where
-	// the keys in the secret/configmap have changed such that some keys have been removed, which would translate
-	// in files having to be removed. commonOSRemove will handle shares mounted off of shares.  And a reminder,
-	// currently this driver does not support overlaying over directories with files.  Either the directory in the
-	// container image must be empty, or the directory does not exist, and is created for the Pod's container as
-	// part of provisioning the container.
-	if err := commonOSRemove(podPath, fmt.Sprintf("commonUpsertRanger key %s volid %s share id %s pod name %s", key, dv.GetVolID(), dv.GetSharedDataId(), dv.GetPodName())); err != nil {
+	// NOTE: atomic_writer handles any pruning of secret/configmap keys where were present before, but are no longer
+	// present
+	if err := os.MkdirAll(podPath, os.ModePerm); err != nil {
 		return err
 	}
-	if err := os.MkdirAll(podPath, os.ModePerm); err != nil {
+	podFile := map[string]atomic.FileProjection{}
+	aw, err := atomic.NewAtomicWriter(podPath, "shared-resource-csi-driver")
+	if err != nil {
 		return err
 	}
 	if payload.ByteData != nil {
```

Comment:
Can a `payload` have both `ByteData` and `StringData` populated at the same time? Can `payload` have both be `nil` also?

Reply:
yes and yes

see the ConfigMap godoc wrt both populated 

both nil means an empty secret/configmap